src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/StringContentAnalysis/StringContentAbstractValue.cs

```diff
@@ -0,0 +1,244 @@
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using Analyzer.Utilities;
+
+namespace Microsoft.CodeAnalysis.Operations.DataFlow.StringContentAnalysis
+{
+    /// <summary>
+    /// Abstract string content data value for symbol/operation tracked by <see cref="StringContentAnalysis"/>.
+    /// </summary>
+    internal partial class StringContentAbstractValue : IEquatable<StringContentAbstractValue>
+    {
+        public static readonly StringContentAbstractValue DefaultMaybe = new StringContentAbstractValue();
+        public static readonly StringContentAbstractValue DefaultNo = new StringContentAbstractValue(literalState: StringContainsState.No, nonLiteralState: StringContainsState.No);
+
+        public StringContentAbstractValue(string literal)
+        {
+            LiteralState = StringContainsState.Yes;
+            NonLiteralState = StringContainsState.No;
+            LiteralValues = ImmutableHashSet.Create(literal);
+            NonLiteralValues = ImmutableHashSet<IOperation>.Empty;
+        }
+
+        public StringContentAbstractValue(IOperation nonLiteral)
+        {
+            LiteralState = StringContainsState.No;
+            NonLiteralState = StringContainsState.Yes;
+            LiteralValues = ImmutableHashSet<string>.Empty;
+            NonLiteralValues = ImmutableHashSet.Create(nonLiteral);
+        }
+
+        private StringContentAbstractValue(StringContainsState literalState = StringContainsState.Maybe, StringContainsState nonLiteralState = StringContainsState.Maybe)
+            : this(literalState, nonLiteralState, ImmutableHashSet<string>.Empty, ImmutableHashSet<IOperation>.Empty)
+        {
+            Debug.Assert(literalState == StringContainsState.No || literalState == StringContainsState.Maybe);
+            Debug.Assert(nonLiteralState == StringContainsState.No || nonLiteralState == StringContainsState.Maybe);
+        }
+
+        private StringContentAbstractValue(StringContainsState literalState, StringContainsState nonLiteralState, ImmutableHashSet<string> literalValues, ImmutableHashSet<IOperation> nonLiteralValues)
+        {
+            LiteralState = literalState;
+            NonLiteralState = nonLiteralState;
+            LiteralValues = literalValues;
+            NonLiteralValues = nonLiteralValues;
+        }
+
+        /// <summary>
+        /// Indicates if this string variable contains string literals or not.
+        /// </summary>
+        public StringContainsState LiteralState { get; }
+
+        /// <summary>
+        /// Indicates if this string variable contains non literal string operands or not.
+        /// </summary>
+        public StringContainsState NonLiteralState { get; }
+
+        /// <summary>
+        /// Gets a collection of the string literals that could possibly make up the contents of this string <see cref="Operand"/>.
+        /// </summary>
+        public ImmutableHashSet<string> LiteralValues { get; }
```

Comment:
It would probably convenient if these were ImmutableOrderedSets instead of unordered HashSets.

Reply:
Actually, scratch that.

---
In reply to: [162766847](https://github.com/dotnet/roslyn-analyzers/pull/1541#discussion_r162766847) [](ancestors = 162766847)

Reply:
I have simplified the implementation of this type to just track: (1) Set of possible literal values and (2) Non literal state. Hopefully, the Merge function is also more intuitive now.

---
In reply to: [162766874](https://github.com/dotnet/roslyn-analyzers/pull/1541#discussion_r162766874) [](ancestors = 162766874,162766847)